projects/pgai/pgai/vectorizer/embeddings.py

```diff
@@ -414,3 +469,191 @@ async def _encode(self, documents: list[str]) -> list[list[int]]:
+            implementation.
+        model (str): The name of the Ollama model used for embeddings.
+        base_url (str): The base url used to access the Ollama API.
+        truncate (bool): Truncate input longer than the model's context length
+        options (dict): Additional ollama-specific runtime options
+        keep_alive (str): How long to keep the model loaded after the request
+    """
+
+    implementation: Literal["ollama"]
+    model: str
+    base_url: str | None = None
+    truncate: bool = True
+    options: OllamaOptions | None = None
+    keep_alive: str | None = None  # this is only `str` because of the SQL API
+
+    @override
+    async def embed(
+        self, documents: list[str]
+    ) -> Sequence[EmbeddingVector | ChunkEmbeddingError]:
+        """
+        Embeds a list of documents into vectors using Ollama's embeddings API.
+
+        If a request to generate embeddings fails because one or more chunks
+        exceed the model's token limit (and truncate is set to False), every
+        chunk will be retried individually. The returned result will contain a
+        ChunkEmbeddingError in place of an EmbeddingVector for the chunks that
+        exceeded the model's token limit.
+
+        Args:
+            documents (list[str]): A list of documents to be embedded.
+
+        Returns:
+            Sequence[EmbeddingVector | ChunkEmbeddingError]: The embeddings or
+            errors for each document.
+        """
+        await logger.adebug(f"Chunks produced: {len(documents)}")
+        try:
+            return await self._batcher.batch_chunks_and_embed(documents)
+        except ollama.ResponseError as e:
+            if "input length exceeds maximum context length" not in e.error:
+                raise e
+            # Note: We don't attempt tokenizing data sent to Ollama models
+            # because every model has its own tokenizer, and there's no way to
+            # programmatically determine the tokenizer. Without knowing the
+            # token length of a chunk, we can't filter the long ones out (like
+            # we do in the OpenAI integration). Instead, we retry each chunk
+            # individually. It's not ideal, but not too bad because it's
+            # (probably) a local API.
+            context_length = await self._context_length()
+            return await self._fallback_retry_individually(context_length, documents)
+
+    @cached_property
+    def _batcher(self) -> BatchApiCaller[StringDocument]:
+        return BatchApiCaller(self._max_chunks_per_batch(), self.call_embed_api)
+
+    @override
+    def _max_chunks_per_batch(self) -> int:
+        # Note: the chosen default is arbitrary - Ollama doesn't place a limit
+        return int(os.getenv("OLLAMA_MAX_CHUNKS_PER_BATCH", default="2048"))
+
+    async def call_embed_api(self, documents: str | list[str]) -> EmbeddingResponse:
+        response = await ollama.AsyncClient(host=self.base_url).embed(
```

Comment:
Wondering if we should allow other client settings, with special emphasis in allowing custom headers. A use case could be for authenticating the request made to ollama's server (auth proxy in front of the server, etc).